service/dap/server.go

```diff
@@ -2569,6 +2574,48 @@ func (s *Server) resetHandlesForStoppedEvent() {
 	s.exceptionErr = nil
 }
 
+func (s *Server) setRunning(running bool) {
+	s.runningMu.Lock()
+	s.running = running
+	s.runningMu.Unlock()
+}
+
+func (s *Server) isRunning() bool {
+	s.runningMu.Lock()
+	defer s.runningMu.Unlock()
+	return s.running
+}
+
+// halt is a helper function to allow dap to make sure that a halt
+// is processed. It is possible that this halt request will come in while a
+// a breakpoint is being processed. If the debugger decides to resume execution
+// after processing the breakpoint, the halt request would be skipped. Additional
+// synchronization is required between "resume" and "halt" to make sure this does
+// not happen.
+func (s *Server) halt() (*api.DebuggerState, error) {
+	s.haltMu.Lock()
+	defer s.haltMu.Unlock()
+	return s.debugger.Command(&api.DebuggerCommand{Name: api.Halt}, nil)
+}
+
+func (s *Server) resume() (*api.DebuggerState, error) {
+	resumeNotify := make(chan struct{}, 1)
+	s.haltMu.Lock()
```

Comment:
I don't think you need to acquire haltMu anymore, nor do you need to pass a non-nil resumeNotify to continue.

Reply:
In fact what is haltMu protecting now?

Reply:
haltMu is protecting the invariant that if a halt request is issued, the process will either halt or remain halted.

If we do not include this, it is possible we could get a halt request right before continuing execution. If this halt came from a setBreakpoints request, the set breakpoints request would hang trying to change breakpoints while targetMutex is held by the continue command.